### PR TITLE
[core] Fixing Python tests for Python 2.7

### PR DESCRIPTION
**Pull request by maceligueta**
The PR is dedicated to make the core Python2.7-compatible.
I was not capable to fix all of them. Everyone is welcome to contribute.
I will be posting the aspect of the errors I am getting, in case you the readers do not have Python 2.7 installed...

### CODE REVIEW

**Comment by maceligueta** [comment]
4 errors left in the Kratos Core:

```
======================================================================
ERROR: test_stream_serialization (test_kratos_parameters.TestParameters)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/maceli/kratos/kratos/tests/test_kratos_parameters.py", line 664, in test_stream_serialization
    serializer = pickle.loads(pickled_data)
  File "/usr/lib64/python2.7/pickle.py", line 1388, in loads
    return Unpickler(file).load()
  File "/usr/lib64/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/usr/lib64/python2.7/pickle.py", line 1089, in load_newobj
    obj = cls.__new__(cls, *args)
TypeError: pybind11_object.__new__(Kratos.StreamSerializer) is not safe, use object.__new__()

======================================================================
ERROR: test_model_serialization_with_pickling (test_model.TestModel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/maceli/kratos/kratos/tests/test_model.py", line 109, in test_model_serialization_with_pickling
    serializer = pickle.loads(pickled_data)
  File "/usr/lib64/python2.7/pickle.py", line 1388, in loads
    return Unpickler(file).load()
  File "/usr/lib64/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/usr/lib64/python2.7/pickle.py", line 1089, in load_newobj
    obj = cls.__new__(cls, *args)
TypeError: pybind11_object.__new__(Kratos.StreamSerializer) is not safe, use object.__new__()

======================================================================
ERROR: test_ComputeNodesMeanNormalModelPart (test_mortar_utilities.TestMortarUtilities)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/maceli/kratos/kratos/tests/test_mortar_utilities.py", line 26, in test_ComputeNodesMeanNormalModelPart
    normal_compute.ComputeNodesMeanNormalModelPart(model_part)
TypeError: unbound method ComputeNodesMeanNormalModelPart() must be called with MortarUtilities instance as first argument (got ModelPart instance instead)

======================================================================
ERROR: test_InvertNormal (test_mortar_utilities.TestMortarUtilities)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/maceli/kratos/kratos/tests/test_mortar_utilities.py", line 55, in test_InvertNormal
    normal_compute.InvertNormal(model_part.Conditions)
TypeError: unbound method InvertNormal() must be called with MortarUtilities instance as first argument (got ConditionsArray instance instead)

----------------------------------------------------------------------
Ran 219 tests in 5.097s

FAILED (errors=4, skipped=20, expected failures=2)

```

**Comment by maceligueta** [comment]
'MortarUtilities' makes me reference @loumalouomega .

**Comment by loumalouomega** [comment]
> 'MortarUtilities' makes me reference @loumalouomega .

What is the problem ?

**Comment by loumalouomega** [comment]
I assume the pybind11 works different on python2 and python3

**Comment by maceligueta** [comment]
ping @loumalouomega

**Comment by pooyan-dadvand** [comment]
@loumalouomega This is very important for the release, would you please fix it?

**Comment by philbucher** [comment]
the serialization-tests are from @RiccardoRossi

**Comment by loumalouomega** [comment]
> @loumalouomega This is very important for the release, would you please fix it?

OK, but yesterday @maceligueta and I we tried and it requires a lot of refactoring for some weird reasons. Today I cannot do it

File: kratos/tests/test_array_1d_interface.py [hunk 1]
```diff
@@ -3,7 +3,12 @@
 
 import KratosMultiphysics.KratosUnittest as KratosUnittest
 
+import sys
+
 class TestArray1DInterface(KratosUnittest.TestCase):
+    def setUp(self):
+        if (sys.version_info < (3, 2)):
+            self.assertRaisesRegex = self.assertRaisesRegexp
```
Comment (philbucher) [inline kratos/tests/test_array_1d_interface.py:11]: @maceligueta I just checked, I think if you add this to the `__init__` of `KratosUnittest.TestCase` it should be fixed globally for all tests.
Could you try please?

**Comment by roigcarlo** [comment]
All errors except the picke one should be solved now. Wonder how the hell this works in python3...
Also, I've created an utility to extract the application name of failed imports with a more generic regex, as some implementation of python may only print the rightmost name of the module.

File: kratos/tests/test_linear_master_slave_constraints.py [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
     import KratosMultiphysics.StructuralMechanicsApplication
     missing_external_dependencies = False
     missing_application = ''
 except ImportError as e:
     missing_external_dependencies = True
     # extract name of the missing application from the error message
-    import re
-    missing_application = re.search(r'''.*'KratosMultiphysics\.(.*)'.*''',
-                                    '{0}'.format(e)).group(1)
+    missing_application = KratosUtils.HandleInvalidImportException(e)
```
Comment (philbucher) [inline kratos/tests/test_linear_master_slave_constraints.py]: @roigcarlo for this I started to use IsApplivationAvailable or AreApplicationsAvailable from kratos_utilities.
Seemed easier to me, what do you think?

Comment (roigcarlo) [reply kratos/tests/test_linear_master_slave_constraints.py]: I'll take a look :+1:

Comment (philbucher) [reply kratos/tests/test_linear_master_slave_constraints.py]: Here is an example:
https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MeshMovingApplication/tests/test_ale_fluid_solver.py

**Comment by roigcarlo** [comment]
@philbucher I've used the functions in utils and adapted the tests. This way is much clear thx!!

@maceligueta, @RiccardoRossi, @loumalouomega For the pickle errors.... Seems that there is a very intersting casuistic here:

We have
- 2 modules: Pickle + cickle.
- 2 versions of python.
- 5 dump Protocols.

Based on the doc from pickle and pybind11:
1 - Python2 + pickle -> not possible
2 - Python2 + cickle -> only protocol 2
3 - Python3 + pickle -> protocols 2 3 or 4
4 - Python3 + cickle -> protocols 2 3 or 4

Based on that I refactored the way the pickle module is imported (cickle preferred, if not possible pickle only if python3)

And added a test skip in case we cannot satisfy 2, 3 or 4.

Hope this solves everything

**Comment by roigcarlo** [comment]
@loumalouomega ping

**Review by loumalouomega — APPROVED** [review]
Approving

**Comment by loumalouomega** [comment]
Merging so we can block the branch

**Pull request merged**